### PR TITLE
Async X-ray computation

### PR DESCRIPTION
**Pull request by sbelak**
Still needs
- [x] API endpoints
- [x] tests
- [x] docs
- [x] canceling
- [x] autocleanup
- [ ] ~~use background threads where available~~
- [x] error handling

### CODE REVIEW

**Comment by sbelak** [comment]
API is provisional (done for table xray) so that we can do a FE proof of concept. I'll add the rest when we settle on the final shape.

File: src/metabase/api/x_ray.clj [hunk 1]
```diff
@@ -56,6 +58,26 @@
                                                  max_computation_cost)})
        fe/x-ray))
 
+(api/defendpoint GET "/async/table/:id"
```
Comment (salsakran) [inline src/metabase/api/x_ray.clj]: this should probably be called `/result/table/:id` or something similar

File: src/metabase/api/x_ray.clj [hunk 2]
```diff
@@ -56,6 +58,26 @@
...
+  "Get x-ray for a `Tield` with ID."
```
Comment (salsakran) [inline src/metabase/api/x_ray.clj]: what's a tield? ;)

**Review by salsakran — COMMENTED** [review]
Some thoughts on frontend concerns:

1) we should probably have a /xray/ComputeJobID URL . Open question whether we should have specific types there. If we do this, then providing x-rays on ad hoc queries (i.e. query dictionaries) becomes easy as we don't need to serialize the dictionary, let alone serialize two for comparisons

2) We should start with a dirt simple polling every X seconds.

File: src/metabase/api/x_ray.clj (hunk 1 above)
Comment (sbelak) [reply src/metabase/api/x_ray.clj]: That's not the result but submitting a computation. You get result via GET /async/:id (I don't think there is much point in having multiple result fetching endpoints)

File: resources/migrations/000_migrations.yaml [hunk 3]
```diff
@@ -3829,3 +3829,87 @@ databaseChangeLog:
...
+                  type: DATETIME
+                  constraints:
+                    nullable: false
+              - column:
+                  name: updated_at
+                  type: DATETIME
+                  constraints:
+                    nullable: false
+              - column:
+                  name: permanence
```
[the remark this answers is not in the document]
Comment (sbelak) [reply resources/migrations/000_migrations.yaml]: I was thinking we might want to keep different results around for different amounts of time, some even save permanently. It's all treated the same currently though.

**Comment by attekei** [comment]
My verbose preparatory notes:

### How computation loading state is currently managed

#### Metadata
* In an unified way with `fetchData` method which dispatches `SET_REQUEST_STATE` action appropriately. 
* The state whether the request has been completed or not can be accessed in Redux state tree through  `state.requests.fetched.metadata[entity_type]` 
* The completion state for a specific type of entity doesn’t exist until it’s being fetched for the first time
* Practically no notion of whether the data is currently (re)loading or not

#### Elsewhere in the app
* Loading state exists in local state trees and usually defaults to `false`
* An action sets the loading state to true and triggers the computation

### How a result of computation is currently managed

#### Metadata
* Stored to central location in Redux state `state.metadata[entity_type]` 
* Accessed via the fully connected metadata graph (which contains empty entity arrays even if they haven’t been loaded yet)

#### Elsewhere in the app
* Computation result doesn’t  exist in state tree until loading is completed (you can use that to infer if we are loading data for the first time)
* There is often  `initialize()` or similar for a specific section of app that resets the state tree and removes computation results so that React components don’t barf because of stale data

### Proposal
* We should create an API that can easily be used for globally shared computations (like metadata) and for more local/scoped computations (like x-rays)
* There should be explicit and separate notions of /whether data is (re)loading/ and /whether we already have data/ and React containers should be able to subscribe to changes in those
* The API difference between computations done with a single http request vs. ”async computations” should be minimal; not sure if  `services.js` should do some abstraction around async requests (maybe not yet)

#### Idea: Letting user to leave and re-enter page while a computation is in progress (not part of MVP)
* Could mean saving the computation request state to localStorage/ sessionStorage
* If there is an existing state available, then don’t trigger a new async computation but use the existing one

#### Idea: Frontend caching of computation results (not part of MVP)
* Something @salsakran played with in [Make selectors great again by salsakran · Pull Request #6017 · metabase/metabase · GitHub](https://github.com/metabase/metabase/pull/6017/files)
* Would improve performance if you want to
	* go back to results of a previous computation using the browser back button (in-memory cache is enough for this)
	* go back to results of some previous computation after quitting the previous browser session (requires localstorage)
* Backend already caches computation results for serving the results quickly so not sure how useful the frontend cache would be atm

**Comment by attekei** [comment]
Implemented an experimental request API that integrates to the local Redux state of app sections and abstracts neatly the difference between "old-school" API endpoints and this new background job / asynchronous computation approach.

You can see the changes I made pretty neatly here:
https://github.com/metabase/metabase/compare/5d3dba236fdd8c0ebf77c8d9baab7e44bf6b0d74...75e4c01031d905ac1f823ded44ea0f3728c898ee

**Comment by attekei** [comment]
It's worth noting that page loading is still blocked by the computations; if we want to change that, we should discuss what to show while the computations are in progress.

**Comment by attekei** [comment]
And something that is even easier to digest for getting an understanding how the API looks like:
```
// What follows is usage of RestfulRequest and BackgroundJobRequest which
// are interchangeable in terms of the method interface of the resulting object

// RestfulRequest for conventional REST API endpoints
// const tableXrayRequest = new RestfulRequest({
//     endpoint: XRayApi.table_xray,
//     resultPropName: 'xray',
//     actionPrefix: 'metabase/xray/table'
// })

// BackgroundJobRequest for
const tableXrayRequest = new BackgroundJobRequest({
    creationEndpoint: XRayApi.async_table_xray,
    statusEndpoint: XRayApi.async_status,
    resultPropName: 'xray',
    actionPrefix: 'metabase/xray/table'
})

export const FETCH_TABLE_XRAY = 'metabase/xray/FETCH_TABLE_XRAY'
export const fetchTableXray = createThunkAction(FETCH_TABLE_XRAY, (tableId, cost) =>
    async (dispatch) => {
        try {
            await dispatch(tableXrayRequest.trigger({ tableId, ...cost.method }))
        } catch (error) {
            console.error(error)
        }
    }
)

export default handleActions({
    ...tableXrayRequest.getReducers(),
    <other reducers here>
}, tableXrayRequest.getDefaultState())
```

**Comment by attekei** [comment]
Now there is a simple caching implementation too because I felt that this PR doesn't really make sense without it. It's currently frontend-only (it uses `localStorage` for saving the job IDs) and we probably can later improve it so that it syncs with the job list retrieved from the backend.

It still takes surprisingly long to load the page even if we are using precomputed data – seems that frontend scripts / rendering takes a long time. I'll audit the page load and see if there are some bottlenecks that could be optimized.

**Comment by sbelak** [comment]
* Moved the entire x-rays api to async.
* Changed async endpoint to be
GET /api/async/:job_id
* added endpoint
GET /api/async/running-jobs 
which returns a vector of ids of all still running jobs belonging to current user

**Comment by salsakran** [comment]
Is there a reason why the BackgroundRequest is created once and re-used?

**Comment by attekei** [comment]
@salsakran Because BackgroundRequest has `getReducers()` and `getDefaultState()` methods that are for constructing the Redux state, see end of the code example in this comment: https://github.com/metabase/metabase/pull/5993#issuecomment-331884546

**Comment by attekei** [comment]
Will throw my thoughts about async job caching here now because my brains still seem to function properly for a few more minutes before they shut down! 💡

Some general questions:
- Should the caching be in backend or frontend or should it be a combination of those two?
- If the caching is in backend, do we need so many endpoints (job creation, job status, job list) or could we eliminate some of those? For instance, should `GET /api/x-ray/table/1` immediately return the cached result if it is available or the ID of an ongoing computation job (instead of creating a new async job)?
- How long should the cached values exist? Is it decided per-endpoint basis (= set in backend) / per frontend use-case basis (= set in frontend) or how?
- Should frontend somehow communicate that "you are now seeing a X minutes/hours/days old xray" ?
- Should frontend let the user to clear the old cached computation value and trigger a new computation?

My observations of the frontend caching POC I created:
- It's definitely nice to be able to jump back and forth between different xrays so that they load instantly (this is more of a general observation than anything frontend caching -specific).
- It's maybe a little dumb that cached values are attached to a specific user browser session. If a colleague of a user wants to see xrays for a table, why wouldn't he see the same cached results? It would be potentially odd if different users see different computation results.

**Comment by salsakran** [comment]
IMO caching should _definitely_ be at least on the backend, precisely because with an async model with result storage, we effectively get it for free. 

Frontend caching is definitely something we can add, though consistency issues are a bit tricky.

With respect to how long they stay cached - Imo, unless someone has better ideas, we should just re-use the cache settings for cards, eg a multiplier on how long they take to compute.

There should _definitely_ be a timestamp telling the user how stale something is.

**Comment by attekei** [comment]
I converted the remaining API endpoints and corresponding Redux actions. Most of xrays should behave normally now.

Notes:
- the experimental frontend caching is still there (I can remove parts related to it very easily when desired, for instance if we want to merge something soon)
- tests fail for some reason, I'll fix those tomorrow
- comparison endpoint for comparing segments and tables doesn't seem to exist atm

**Comment by sbelak** [comment]
Added :created-at to job return, so we can communicate how stale the result is. 

Shall we add BE caching to this PR, or is it going to be too many moving parts?

The only thing that's not clear to me at this point re:caching is should we do it on a per user+endpoint level or just per endpoint? Two users could potentially have different (row-level) permissions.

**Comment by attekei** [comment]
I removed the frontend caching logic so we could merge this PR already if we want (and address backend caching in a separate PR). But I don't have a strong preference.

File: src/metabase/feature_extraction/async.clj [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  (when (running? job)
+    (future-cancel (@running-jobs id))
+    (swap! running-jobs dissoc id)
+    (db/update! ComputationJob id :status :canceled)))
+
+(defn compute
+  "Compute closure `f` asynchronously. Returns id of the associated computation
+   job."
+  [f]
+  (let [job (db/insert! ComputationJob
```
Comment (salsakran) [inline src/metabase/feature_extraction/async.clj]: You should log something to indicate the job started + stopped, and any errors thrown.

**Review by salsakran — CHANGES_REQUESTED** [review]

**Comment by salsakran** [comment]
Overall this seems to work, but when it doesn't, there's no way to figure out what happened. 
eg 
screen shot 2017-10-02 at 7 09 34 pm
isn't too useful.

I'm a little skittish of the 1s retry, and think we should introduce some sort of back-off there as well.

File: src/metabase/feature_extraction/async.clj (hunk 4 above)
Comment (sbelak) [reply src/metabase/feature_extraction/async.clj]: If the error was in what gets computed, it gets stored same as results (see `save-error`), if it's in the computation framework it 500s same as we do elsewhere. We should catch both Errors and Exceptions though, so things like OutOfMemory below will get caught. Not entirely sure how to handle other Throwables as we do use them for flow control (eg. `api/read-check`).

Comment (salsakran) [reply src/metabase/feature_extraction/async.clj]: We should still log things in the server logs. It's a basic pattern we use all over the place -- if something on the server goes wrong, throw it into the server logs.

Not all administrators or analysts will have access to that session, or the underlying application database.

**Review by salsakran — APPROVED** [review]

**Comment by salsakran** [comment]
tests?

**Comment by sbelak** [comment]
Added endpoint tests, the rest were already there. Miss anything in particular?

**Comment by sbelak** [comment]
Speaking of tests, since I've just added xray api endpoint tests in  GH-6035, the changes in this branch will break them. Ideally we'd first review and merge that one, then I fix the tests here.

**Comment by salsakran** [comment]
segment comparison is broken. 

I get the below error
`GET /api/x-ray/compare/segment/1/table/3 does not exist.`

**Pull request merged**